### DISCUSSION TITLE
Remove assets after testing file transfer

### DISCUSSION THREAD

**Discussion by veronikasiska** [eclipse-edc/Connector#3311] [Q&A] [answered]
Dear Community, I have some trouble with implementing a clean-up for a file transfer end-to-end test. The initial idea was to clean up all resources (assets, policies, contract definitions) using EDC, but the asset can not be removed since it is associated with a contract agreement. I do not see a way to remove a contract agreement via the management API.

Am I missing something? Or is it intended that once an asset has been part of a transfer, it can not be removed any more?

Many thanks in advance for the help!

**Comment by ndr-brt** [comment, marked as answer]
it is intended, a Contract Agreement is a fact that happened and currently there's no way to delete one.

**Comment by veronikasiska** [comment]
Thank you for the answer! That part is clear (that the contract agreement can not be deleted). Still, is it intended that once an asset has a contract agreement associated with it, it can never ever be deleted again?

**Reply by ndr-brt** [reply]
It's about consistency, as you said the asset is associated with the agreement, so deleting it would cause data inconsistency.

**Reply by veronikasiska** [reply]
That's perfectly understandable, but does not seem very realistic from the perspective of a provider. I expect to be able to remove offers and assets, even if they were subject of valid contracts before. There could be constraints to ensure that data consistency is retained, but I can imagine cases where this could be done in a consistent way; e.g. the policies attached to all contract agreements for that asset have an expiration date that lies in the past.

What do you think? Has this been discussed before?

**Reply by jimmarino** [reply]
Yes, you can implement a collector that transitively determines which assets no longer reference in-force contracts and remove them. Since this is use-case specific, it should be handled as a customization.
